perfaware/sim86/shared/Sim8086.cpp

```diff
@@ -0,0 +1,162 @@
+#include <iostream>
+#include <fstream>
+#include <vector>
+#include <string>
+#include <sstream>
+#include "sim86_shared.h"
+#pragma comment (lib, "sim86_shared_debug.lib")
+#define ArrayCount(Array) (sizeof(Array) / sizeof((Array)[0]))
+
+s16 Registers[15] = {};
+
+char const* RegisterNames[][3] =
+{
+    {"", "", ""},
+    {"al", "ah", "ax"},
+    {"bl", "bh", "bx"},
+    {"cl", "ch", "cx"},
+    {"dl", "dh", "dx"},
+    {"sp", "sp", "sp"},
+    {"bp", "bp", "bp"},
+    {"si", "si", "si"},
+    {"di", "di", "di"},
+    {"es", "es", "es"},
+    {"cs", "cs", "cs"},
+    {"ss", "ss", "ss"},
+    {"ds", "ds", "ds"},
+    {"ip", "ip", "ip"},
+    {"flags", "flags", "flags"}
+};
+
+void SimulateInstruction(instruction Instruction)
+{
+    const char* Op = Sim86_MnemonicFromOperationType(Instruction.Op);
+    
+    if (Instruction.Op == Op_mov)
+    {
+        instruction_operand Dest = Instruction.Operands[0];
+        instruction_operand Source = Instruction.Operands[1];
+        s32 Value; // s32 to avoid issues on the bitshift right
+        switch (Source.Type)
+        {
+            case Operand_Register:
+            {
+                Value = Registers[Source.Register.Index];
+                if (Source.Register.Offset == 1) // Accessing high 8 bits
+                {
+                    Value >>= 8;
+                }
+                else if (Source.Register.Count == 1) // Accessing low 8 bits
+                {
+                    Value &= 0xFF;
+                }
+            } break;
+            case Operand_Immediate:
+            {
+                Value = Source.Immediate.Value;
+            } break;
+        }
+
+        Value = (s16)Value;
+
+        switch (Dest.Type)
+        {
+            case Operand_Register:
+            {
+                if (Dest.Register.Offset == 1) // Writing high 8 bits
+                {
+                    s16 Temp = Registers[Dest.Register.Index];
+                    Temp &= 0xFF;
+                    Value <<= 8;
+                    Registers[Dest.Register.Index] = Temp | Value;
+                }
+                else if (Dest.Register.Count == 1) // Accessing low 8 bits
+                {
+                    Value &= 0xFF;
+                    s16 Temp = Registers[Dest.Register.Index];
+                    Temp &= 0xFF00;
+                    Registers[Dest.Register.Index] = Temp | Value;
+                }
```

Comment:
I think that's fine, although maybe you don't need the `Temp` and can do it all directly on the register.

But here's what I would have done, which I think is more explicit about the intent and easier to read, and doesn't require to handle low and high bits separately:
```
if (Dest.Register.Count == 1)
{
    u8 *Dest = static_cast<u8 *>(&Registers[Dest.Register.Index]) + Dest.Register.Offset;
    *Dest = static_cast<u8>(Value);
}
```

Reply:
Fixed here 9bf771ca8ef7a0d2a91e5ee9f9da53f72d0a4a8f
A straight static_cast to <u8*> didn't work. I don't understand it completely but there are restrictions between pointer type casts. 
See here: https://en.cppreference.com/w/cpp/language/static_cast

> Two objects a and b are pointer-interconvertible if:
    they are the same object, or
    one is a union object and the other is a non-static data member of that object, or
    one is a standard-layout class object and the other is the first non-static data member of that object or any base class subobject of that object, or
    there exists an object c such that a and c are pointer-interconvertible, and c and b are pointer-interconvertible. 

I found 2 options:
- static_cast to void* then to u8*
- reinterpret_cast straight to u8*
I used the first one because it seemed safer to me, but I think reinterpret_cast would remain safe as it would always be a downcast.

Reply:
You can use `reinterpret_cast` yes. Also regarding your last comment, `static_cast` shouldn't be generating any instructions here neither, it only does so when there is an operation to perform on the data (sign extension, or whatever). For pointers it doesn't do any operation, but it does some checking at compile time for safety. Here it doesn't work for the reasons you pointed out, but `reinterpret_cast` will work indeed.

Reply:
Actually now that I think about it it might do some operation on pointers. This would be the case if you cast from a derived class to a base class, then it might apply an offest (maybe because of the pointer to account for the virtual table pointer which compilers often place as the first thing in the memory for the object, or maybe because you inherit from multiple classes and you cast to one which is not the first one, etc).

Here is an example, see the asm corresponding to the cast lines to see that `static_cast` did indeed perform operations: https://godbolt.org/z/dxEqWoKrb

Reply:
Awesome, thanks for showing an example!
Fixed c349af57aff1602572e88b50d9d2248e4c03821d